[DO NOT MERGE] Sim changes for forthcoming beta MicroPython release

microbit-matt-hillsdon opened the pull request.

microbit-matt-hillsdon commented on src/board/audio/index.ts:
```diff
@@ -155,7 +156,80 @@ export class Audio {
+    source.connect(gain);
+    // TODO: consider AudioWorklet - worth it? Browser support?
+    //       consider alternative resampling approaches
+    //       what sample rates are actually supported this way?
```
[ScriptProcessorNode support](https://caniuse.com/mdn-api_scriptprocessornode) vs [AudioWorklet support](https://caniuse.com/mdn-api_audioworklet) so we'd lose early Safari 14.x and Safari 13 (for 13 I think limited support is OK).

See also the techniques used [in this project](https://github.com/chrisguttandin/extendable-media-recorder/tree/f0badc8da7ba26c18980970ea8ac91206bfec7cd).

microbit-matt-hillsdon commented on src/board/pins.ts:
```diff
@@ -128,8 +136,21 @@ export class TouchPin extends BasePin {
+  getAndClearTouches() {
+    const touches = this._touches;
+    this._touches = 0;
+    console.log("got touch ", touches);
```
Remove debug

```suggestion
```

microbit-matt-hillsdon commented on src/board/wasm.ts:
```diff
@@ -9,13 +9,14 @@ export interface EmscriptenModule {
   _microbit_radio_rx_buffer(): number;
 
   HEAPU8: Uint8Array;
+  HEAPU32: Uint32Array;
```
Unused in the end but we also need to find/add type definitions for Emscripten's setValue/getValue pointer API.

microbit-matt-hillsdon commented on src/mphalport.h:
```diff
@@ -9,6 +9,8 @@
 
 extern ringbuf_t stdin_ringbuf;
 
+#define mp_hal_ticks_cpu() (0)
```
Need to understand this / investigate new time API.
We need to do the stubs for that too.

microbit-matt-hillsdon replied:
Some discussion on the stubs PR too

jaustin commented:
Having used this a little, I've noticed the microphone LED isn't coming on when using the computer mic.

microbit-matt-hillsdon commented:
Re the Firefox regression the sample rate seems to have changed.

Main, working:
![image](https://github.com/microbit-foundation/micropython-microbit-v2-simulator/assets/44397098/9c760fbb-0b10-4907-81d7-9084d4c8e4ad)

Branch, not working:
![image](https://github.com/microbit-foundation/micropython-microbit-v2-simulator/assets/44397098/fd205468-a803-4660-9479-42b0deb2e7f4)

This 4x reduction in sample rate is due to the removal of buffer expansion in 2c67ed68a15bb64949e7825eaba46acb57f680b2 which drops the sample rate out of Firefox's supported range (older Safari will have issues too as we saw previously for speech).

I guess we need to look at resampling in the sim for multiple reasons: 
1. increasing the audio frame playback rate 
2. downsampling the media recorder rate so we can get data at the requested rate

microbit-matt-hillsdon commented:
Picking this up again.

MicroPython changes since the last update for reference:
https://github.com/microbit-foundation/micropython-microbit-v2/compare/17aaa115ad4ad285377b2eb210f37b0e94f61aa9..audio-recording#diff-3d91b87d622cba0397eed7b4ff512602fa464e81ad11916d4b26722e5b0162af
